### ISSUE TITLE
Make JsEnv extensible

### ISSUE THREAD

**Issue by nafg** [com-lihaoyi/mill#2144]
In Mill, the JsEnv selection is based on a sealed trait with only 3 choices: NodeJs, JsDom, and Phantom (which is an obsolete project). But there are many cases when one would want to use a different strategy. As far as I can tell, currently the only way to accomplish that now is to override `testTask` and do _everything_ from scratch, because all of the scalajs APIs are private.

One use case for this is described in https://github.com/scala-js/scala-js-env-jsdom-nodejs/issues/44, where @sjrd suggests using https://github.com/exoego/scala-js-env-jsdom-nodejs. Hopefully that project will continue to be maintained, but in any case it's out of reach from Mill, as far as I can tell.

I think it would need a major redesign of the APIs. IIUC (mentioned in another issue) the reason for the current design is to enable one project to use multiple versions of Scala.js, which is a great feature that even SBT doesn't support. I don't know if it's possible to have both at the same time (user-supplied JsEnv and multiple scalajs versions in one project).

But I don't see any other way to use jsdom to test code that uses modules.

**Comment by sjrd** [comment]
> IIUC (mentioned in another issue) the reason for the current design is to enable one project to use multiple versions of Scala.js, which is a great feature that even SBT doesn't support. I don't know if it's possible to have both at the same time (user-supplied JsEnv and multiple scalajs versions in one project).

Since Scala.js 1.1.1, the `JSEnv` APIs come from separate artifacts with their own, backwards binary compatible versioning. So Mill could always use the latest version of `scalajs-js-envs` without compromising its ability to link with multiple versions of Scala.js.

**Comment by lefou** [comment]
I think some internal used code is private to avoid the maintenance costs of preserving binary compatibility, which we take seriously in Mill. But it's not set in stone, so if you come up with a good use case and a proper suggestion which API you need to change and how, we can change / improve it.

You need to be a bit more specific though, as your current description isn't actionable for me. 

Also, I admit that Scala.js is not my field of expertise, so you better be verbose when describing the needed changes. I would prefer some draft pull request, so we can focus on concrete code and the impact of changes in terms of tests.

That being said, I'm very open to support more use cases and looking forward to it.

**Comment by lolgab** [comment]
@sjrd I tried in the past to use the Scala.js API directly, but when changing version of Scala.js it was breaking the classpath. Maybe if we compiled with Scala.js `1.0.0` but then we couldn't use the newer features. If we compile against the latest versions then I remember that the older versions were breaking.
I think it is a similar problem to the one faced in mdoc https://github.com/scalameta/mdoc/pull/629 which was solved by creating a copy of the Scala.js API and mapping it to the real Scala.js types in the separate classpath.
Since we need support every version at the same time the current approach is probably the best we can have.
